archive/imo/imo2013_q5.lean

```diff
@@ -0,0 +1,423 @@
+  rw [sub_add_cancel x y, geom_series₂_def] at this,
+  nlinarith,
+end
+
+lemma le_of_all_pow_lt_succ (x y : ℝ) (hx : 1 < x) (hy : 1 < y)
+      (h : ∀n:ℕ, 0 < n → x^n - 1 < y^n)
+      : x ≤ y :=
+begin
+   by_contra hxy,
+   push_neg at hxy,
+   have hxmy : 0 < x - y := sub_pos.mpr hxy,
+   have hn: (∀n:ℕ, 0 < n → (x - y) * (n:ℝ) ≤ x^n - y^n),
+   {
+     intros n hn,
+     have hterm : (∀i:ℕ, i ∈ finset.range n → 1 ≤ x^i * y^(n - 1 - i)),
+     {
+       intros i hi,
+       have hx' : 1 ≤ x ^ i := one_le_pow_of_one_le (le_of_lt hx) i,
+       have hy' : 1 ≤ y ^ (n - 1 - i) := one_le_pow_of_one_le (le_of_lt hy) (n - 1 - i),
+       nlinarith,
+     },
+     have : ((n:ℝ) = (∑ (i : ℕ) in finset.range n, (1:ℝ))),
+     {
+       simp only [mul_one, finset.sum_const, nsmul_eq_mul, finset.card_range]
+     },
+
+     calc (x - y) * (n:ℝ)
+             = (x - y) *  (∑ (i : ℕ) in finset.range n, (1:ℝ))
+               : by simp only [mul_one, finset.sum_const, nsmul_eq_mul, finset.card_range]
+         ... ≤ (x-y) * (∑ (i : ℕ) in finset.range n, x ^ i * y ^ (n - 1 - i))
+               : (mul_le_mul_left hxmy).mpr (finset.sum_le_sum hterm)
+         ... = x^n - y^n : (factor_xn_minus_yn x y n).symm,
+   },
+
+   -- Choose n larger than 1 / (x - y).
+   obtain ⟨N, hN⟩ := exists_nat_gt (1 / (x - y)),
+   have hNpr : (0:ℝ) < N := lt_trans (one_div_pos.mpr hxmy) hN,
+   have hNp : 0 < N,
+   {
+      cases N,
+      { exfalso, exact lt_asymm hNpr hNpr },
+      exact nat.succ_pos N
+   },
+
+   have := calc 1 = (x - y) * (1 / (x - y)) : by field_simp [ne_of_gt hxmy]
+              ... < (x - y) * N             : (mul_lt_mul_left hxmy).mpr hN
+              ... ≤ x^N - y^N               : hn N hNp,
+   linarith [h N hNp],
+end
+
+/--
+ Like le_of_all_pow_lt_succ, but with a weaker assumption for y.
+-/
+lemma le_of_all_pow_lt_succ' (x y: ℝ) (hx: 1 < x) (hy: 0 < y)
+      (h: ∀n:ℕ, 0 < n → x^n - 1 < y^n)
+      : x ≤ y :=
+begin
+  have hy': 1 < y,
+  {
+    by_contra hy'',
+    push_neg at hy'', -- hy'' :  y ≤ 1.
+
+    -- Then there exists y' such that 0 < y ≤ 1 < y' < x.
+    let y' := (x + 1) / 2,
+    have h_y'_lt_x: y' < x,
+    {
+      have hh: (x + 1)/2 < (x * 2) / 2 := by linarith,
+      calc y' < (x * 2) / 2 : hh
+          ... = x : by field_simp
+    },
+    have h1_lt_y' : 1 < y',
+    {
+      have hh': 1 * 2 / 2 < (x + 1) / 2 := by linarith,
+      calc 1 = 1 * 2 / 2 : by field_simp
+         ... < y' : hh'
+    },
+    have h_y_lt_y' := calc y ≤ 1 : hy''
+                         ... < y': h1_lt_y',
+    have hh: (∀ n, 0 < n → x^n - 1 < y'^n),
+    {
+      intros n hn,
+      calc x^n - 1 < y^n : h n hn
+              ...  ≤ y'^n : pow_le_pow_of_le_left (le_of_lt hy) (le_of_lt h_y_lt_y') n
+    },
+    have : x ≤ y' := le_of_all_pow_lt_succ x y' hx h1_lt_y' hh,
+    linarith, -- contradiction
+  },
+  exact le_of_all_pow_lt_succ x y hx hy' h
+end
+
+lemma power_bound (n: ℕ) (ε:ℚ) (ha: 0 < ε) : 1 + (n:ℚ) * ε ≤ (1 + ε)^n :=
```

Comment:
This seems like a more generally useful lemma that belongs in mathlib proper. Not in the specific form with `ℚ` and `ℕ`, but in various forms with the exponent being from any one of `ℕ`, `ℤ` or `ℝ` (in the first two cases you can probably prove something for powers of numbers in some kind of ordered semiring / ordered group) and whatever the weakest inequalities on the arguments are for the result to be true.

Reply:
I agree that this could be generally useful. I've pushed a commit that generalizes ℚ to `linear_ordered_comm_ring`, and I've added a TODO about moving it into mathlib proper. I don't know about generalizing the exponent -- that sounds like it might make things a bit more complicated.

Reply:
I know this result as Bernoulli's inequality, and I see that there is one version in mathlib already:
https://leanprover-community.github.io/mathlib_docs/algebra/group_power/lemmas.html#one_add_mul_le_pow'

Reply:
Ah, and actually [one_add_sub_mul_le_pow](https://leanprover-community.github.io/mathlib_docs/algebra/group_power/lemmas.html#one_add_sub_mul_le_pow) is even closer to what I need. Thanks!